### PR TITLE
[rest] improve clusters rest endpoint by documenting the endpoints with more swagger annotations

### PR DESCRIPTION
**Pull request by sijie**

### CODE REVIEW

**Comment by sijie** [comment]
@merlimat @rdhabalia please take a look at this change. I included swagger annotation in `pulsar-common` for documenting the request/responses. Not sure if that is okay. Please share your thoughts.

@Jennifer88huang @Anonymitaet please help with the language part.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/ClustersBase.java [hunk 1]
```diff
@@ -135,11 +186,37 @@ public void createCluster(@PathParam("cluster") String cluster, ClusterData clus
...
-    @ApiResponses(value = { @ApiResponse(code = 204, message = "Cluster has been updated"),
-            @ApiResponse(code = 403, message = "Don't have admin permission"),
-            @ApiResponse(code = 404, message = "Cluster doesn't exist") })
-    public void updateCluster(@PathParam("cluster") String cluster, ClusterData clusterData) {
+    @ApiOperation(
+        value = "Update the configuration for a cluster.",
+        notes = "This operation requires Pulsar super-user privileges.")
+    @ApiResponses(value = {
+            @ApiResponse(code = 204, message = "Cluster has been updated"),
+            @ApiResponse(code = 403, message = "Don't have admin permission or policies are readonly"),
```
Comment (Anonymitaet) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/ClustersBase.java]:
```suggestion
            @ApiResponse(code = 403, message = "Don't have admin permission or policies are read-only"),
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/ClustersBase.java [hunk 2]
```diff
@@ -329,11 +458,24 @@ private void deleteFailureDomain(String clusterPath) {
 
     @GET
     @Path("/{cluster}/namespaceIsolationPolicies")
-    @ApiOperation(value = "Get the namespace isolation policies assigned in the cluster", response = NamespaceIsolationData.class, responseContainer = "Map")
-    @ApiResponses(value = { @ApiResponse(code = 403, message = "Don't have admin permission"),
-            @ApiResponse(code = 404, message = "Cluster doesn't exist") })
-    public Map<String, NamespaceIsolationData> getNamespaceIsolationPolicies(@PathParam("cluster") String cluster)
-            throws Exception {
+    @ApiOperation(
+        value = "Get the namespace isolation policies assigned in the cluster",
```
Comment (Anonymitaet) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/ClustersBase.java]:
```suggestion
        value = "Get the namespace isolation policies assigned to the cluster",
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/ClustersBase.java [hunk 3]
```diff
@@ -354,12 +496,29 @@ private void deleteFailureDomain(String clusterPath) {
 
     @GET
     @Path("/{cluster}/namespaceIsolationPolicies/{policyName}")
-    @ApiOperation(value = "Get a single namespace isolation policy assigned in the cluster", response = NamespaceIsolationData.class)
-    @ApiResponses(value = { @ApiResponse(code = 403, message = "Don't have admin permission"),
+    @ApiOperation(
+            value = "Get a single namespace isolation policy assigned in the cluster",
```
Comment (Anonymitaet) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/ClustersBase.java]:
```suggestion
            value = "Get the single namespace isolation policy assigned to the cluster",
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/ClustersBase.java [hunk 4]
```diff
@@ -354,12 +496,29 @@ private void deleteFailureDomain(String clusterPath) {
...
+            @ApiResponse(code = 500, message = "Internal server error")
+    })
+    public NamespaceIsolationData getNamespaceIsolationPolicy(
+        @ApiParam(
+            value = "The cluster name",
+            required = true
+        )
+        @PathParam("cluster") String cluster,
+        @ApiParam(
+            value = "The namespace isolation policy name",
```
Comment (Anonymitaet) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/ClustersBase.java]:
```suggestion
            value = "The policy name of the namespace isolation",
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/ClustersBase.java [hunk 5]
```diff
@@ -385,11 +544,23 @@ public NamespaceIsolationData getNamespaceIsolationPolicy(@PathParam("cluster")
 
     @GET
     @Path("/{cluster}/namespaceIsolationPolicies/brokers")
-    @ApiOperation(value = "Get list of brokers with namespace-isolation policies attached to them", response = BrokerNamespaceIsolationData.class)
-    @ApiResponses(value = { @ApiResponse(code = 403, message = "Don't have admin permission"),
-            @ApiResponse(code = 404, message = "Namespace-isolation policies not found"),
-            @ApiResponse(code = 412, message = "Cluster doesn't exist") })
+    @ApiOperation(
+        value = "Get list of brokers with namespace-isolation policies attached to them",
```
Comment (Anonymitaet) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/ClustersBase.java]:
```suggestion
        value = "Get the list of brokers with namespace isolation policies attached to them",
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/ClustersBase.java [hunk 6]
```diff
@@ -684,11 +965,28 @@ public FailureDomain getDomain(@PathParam("cluster") String cluster, @PathParam(
 
     @DELETE
     @Path("/{cluster}/failureDomains/{domainName}")
-    @ApiOperation(value = "Delete cluster's failure omain")
-    @ApiResponses(value = { @ApiResponse(code = 403, message = "Don't have admin permission or plicy is read only"),
-            @ApiResponse(code = 412, message = "Cluster doesn't exist") })
-    public void deleteFailureDomain(@PathParam("cluster") String cluster, @PathParam("domainName") String domainName)
-            throws Exception {
+    @ApiOperation(
+        value = "Delete cluster's failure domain",
```
Comment (Anonymitaet) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/ClustersBase.java]:
```suggestion
        value = "Delete the failure domain of the cluster",
```

File: pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/NamespaceIsolationData.java [hunk 7]
```diff
@@ -20,16 +20,45 @@
...
+    )
     public List<String> primary = new ArrayList<String>();
+    @ApiModelProperty(
+        name = "primary",
+        value = "The list of secondary brokers for serving the list of namespaces in this isolation policy"
+    )
     public List<String> secondary = new ArrayList<String>();
+    @ApiModelProperty(
+        name = "auto_failover_policy",
+        value = "The auto failover policy configuration data",
```
Comment (Anonymitaet) [inline pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/NamespaceIsolationData.java]:
```suggestion
        value = "The data of auto-failover policy configuration",
```

File: pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/NamespaceIsolationData.java [hunk 8]
```diff
@@ -20,16 +20,45 @@
...
+import io.swagger.annotations.ApiModel;
+import io.swagger.annotations.ApiModelProperty;
 import java.util.ArrayList;
 import java.util.List;
 
 import com.google.common.base.Objects;
 
+@ApiModel(
+    value = "NamespaceIsolationData",
+    description = "The namespace isolation configuration data"
```
Comment (Anonymitaet) [inline pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/NamespaceIsolationData.java]:
```suggestion
    description = "The data of namespace isolation configuration"
```

File: pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/FailureDomain.java [hunk 9]
```diff
@@ -20,11 +20,22 @@
...
 import com.google.common.base.MoreObjects;
 import com.google.common.base.Objects;
+import io.swagger.annotations.ApiModel;
+import io.swagger.annotations.ApiModelProperty;
 import java.util.HashSet;
 import java.util.Set;
 
+@ApiModel(
+    value = "FailureDomain",
+    description = "The configuration data for a Failure Domain in a cluster"
```
Comment (Anonymitaet) [inline pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/FailureDomain.java]:
```suggestion
    description = "The data of a failure domain configuration in a cluster"
```

File: pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/ClusterData.java [hunk 10]
```diff
@@ -20,18 +20,48 @@
...
 public class ClusterData {
+    @ApiModelProperty(
+        name = "serviceUrl",
+        value = "The http rest service url (for admin operations)",
+        example = "http://pulsar.example.com:8080"
+    )
     private String serviceUrl;
+    @ApiModelProperty(
+        name = "serviceUrlTls",
+        value = "The https rest service url (for admin operations)",
```
Comment (Anonymitaet) [inline pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/ClusterData.java]:
```suggestion
        value = "The https rest service URL (for admin operations)",
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/ClustersBase.java [hunk 11]
```diff
@@ -108,12 +132,39 @@ public ClusterData getCluster(@PathParam("cluster") String cluster) {
 
     @PUT
     @Path("/{cluster}")
-    @ApiOperation(value = "Provisions a new cluster. This operation requires Pulsar super-user privileges.", notes = "The name cannot contain '/' characters.")
-    @ApiResponses(value = { @ApiResponse(code = 204, message = "Cluster has been created"),
+    @ApiOperation(
+        value = "Create a new cluster. ",
+        notes = "This operation requires Pulsar super-user privileges, and the name cannot contain '/' characters."
```
Comment (Anonymitaet) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/ClustersBase.java]:
```suggestion
        notes = "This operation requires Pulsar superuser privileges, and the name cannot contain the " / " character."
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/ClustersBase.java [hunk 12]
```diff
@@ -596,12 +824,34 @@ public void deleteNamespaceIsolationPolicy(@PathParam("cluster") String cluster,
...
     @POST
     @Path("/{cluster}/failureDomains/{domainName}")
-    @ApiOperation(value = "Set cluster's failure Domain")
-    @ApiResponses(value = { @ApiResponse(code = 403, message = "Don't have admin permission"),
-            @ApiResponse(code = 409, message = "Broker already exists in another domain"),
-            @ApiResponse(code = 404, message = "Cluster doesn't exist") })
-    public void setFailureDomain(@PathParam("cluster") String cluster, @PathParam("domainName") String domainName,
-            FailureDomain domain) throws Exception {
+    @ApiOperation(
+        value = "Set cluster's failure Domain",
```
Comment (Anonymitaet) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/ClustersBase.java]:
```suggestion
        value = "Set the failure domain of the cluster",
```

File: pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/ClusterData.java [hunk 13]
```diff
@@ -20,18 +20,48 @@
...
 import com.google.common.base.MoreObjects;
 
+@ApiModel(
+    value = "ClusterData",
+    description = "The configuration data for a cluster"
+)
 public class ClusterData {
+    @ApiModelProperty(
+        name = "serviceUrl",
+        value = "The http rest service url (for admin operations)",
```
Comment (Jennifer88huang) [inline pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/ClusterData.java]: HTTP, URL
When we use the abbreviation form for protocol names or proper nouns in descriptive text, we use the uppercase letters.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/ClustersBase.java [hunk 14]
```diff
@@ -434,12 +605,29 @@ public NamespaceIsolationData getNamespaceIsolationPolicy(@PathParam("cluster")
...
-    public BrokerNamespaceIsolationData getBrokerWithNamespaceIsolationPolicy(@PathParam("cluster") String cluster,
-            @PathParam("broker") String broker) {
+    @ApiOperation(
+        value = "Get a broker with namespace-isolation policies attached to it",
+        response = BrokerNamespaceIsolationData.class,
+        notes = "This operation requires Pulsar super-user privileges."
+    )
+    @ApiResponses(value = {
+        @ApiResponse(code = 403, message = "Don't have admin permission"),
+        @ApiResponse(code = 404, message = "Namespace-isolation policies/ Broker not found"),
```
Comment (Jennifer88huang) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/ClustersBase.java]: Do you mean the error code "404" apply to both of the error messages "Namespace-isolation policies/ Broker not found"?

Comment (sijie) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/ClustersBase.java]: yes in either case, pulsar returns 404

**Comment by srkukarni** [comment]
run java8 tests
run cpp tests

**Comment by sijie** [comment]
Please don't merge this yet. I haven't fully incorporated the review comments from @Jennifer88huang and @Anonymitaet

**Comment by sijie** [comment]
@Jennifer88huang @Anonymitaet I updated this PR. please take a look again.

**Comment by sijie** [comment]
run integration tests

**Pull request merged**